packages/backend-data/src/convert_js_resolvers.ts

```diff
@@ -71,12 +47,38 @@ export const convertJsResolverDefinition = (
 
     const resolverName = `Resolver_${resolver.typeName}_${resolver.fieldName}`;
 
+    const amplifyEnvironmentName =
+      scope.node.tryGetContext('amplifyEnvironmentName') ?? 'NONE';
     new CfnResolver(scope, resolverName, {
       apiId: amplifyApi.apiId,
       fieldName: resolver.fieldName,
       typeName: resolver.typeName,
       kind: APPSYNC_PIPELINE_RESOLVER,
-      codeS3Location: jsResolverTemplateAsset.s3ObjectUrl,
+      /**
+       * The top-level passthrough resolver request/response handler (see: https://docs.aws.amazon.com/appsync/latest/devguide/resolver-reference-overview-js.html#anatomy-of-a-pipeline-resolver-js)
+       * It's required for defining a pipeline resolver. Adds the GraphQL API ID and Amplify environment name to the context stash.
+       * Returns the output of the last function in the pipeline back to the client.
+       *
+       * Customer-provided handlers are added as a Functions list in `pipelineConfig.functions`
+       *
+       * Uses synth-time inline code to avoid circular dependency when adding the API ID as an environment variable.
+       */
+      code: `
+        /**
+         * Pipeline resolver request handler
+         */
+        export const request = (ctx) => {
+          ctx.stash.awsAppsyncApiId = '${amplifyApi.apiId}';
+          ctx.stash.awsAmplifyEnvironmentName = '${amplifyEnvironmentName}';
+          return {};
+        };
+        /**
+         * Pipeline resolver response handler
+         */
+        export const response = (ctx) => {
+          return ctx.prev.result;
+        };
+      `,
```

Comment:
I've changed to use the compiled file. We still have the risk of malformed JS because we need to add in the GraphQL API ID and Amplify environment name after compilation. Let me know if we think we still need a E2E test to cover this.

Reply:
thanks.

> Let me know if we think we still need a E2E test to cover this.

if app id and env injection can be run as part of unit/integration test then e2e is not required. 

i.e. in the unit/integration test we could:
1. run injection logic and dump resulting file on disk.
2. dynamically import it
3. run test scenarios on dynamic import.